circleguard/investigator.py

```diff
@@ -226,43 +250,154 @@ def median_frametime(frametimes):
         return np.median(frametimes)
 
     @staticmethod
-    def _parse_beatmap(beatmap):
-        hitobjs = []
+    def keydown_frames(replay):
+        """
+        Get the frames of ``replay`` which had a keydown event, and we should
+        consider eligible to hit a hitobject.
 
-        # parse hitobj
-        for hit in beatmap.hit_objects_no_spinners:
-            p = hit.position
-            hitobjs.append([hit.time.total_seconds() * 1000, p.x, p.y])
-        return hitobjs
+        Parameters
+        ----------
+        replay: :class:`~.Replay`
+            The replay to get the keydown frames of.
 
-    @staticmethod
-    def _parse_keypress_times(replay):
+        Returns
+        -------
+        ndarray(float, [float, float], bool)
+            The keydown frames for the replay. The first float is the time of
+            that frame, the second and third floats are the x and y position
+            of the cursor at that frame, and the bool is whether this frame was
+            a zero frametime frame.
+        """
+        keydown_frames = []
+        # the keys currently down for each frame
         keypresses = replay.k & Investigator.MASK
-        changes = keypresses & ~np.insert(keypresses[:-1], 0, 0)
-        return replay.t[changes != 0]
 
+        # the keydowns for each frame. Frames are "keydown" frames if an
+        # additional key was pressed from the previous frame. If keys pressed
+        # remained the same or decreased (a key previously pressed is no longer
+        # pressed) from the previous frame, ``keydowns`` is zero for that frame.
+        keydowns = keypresses & ~np.insert(keypresses[:-1], 0, 0)
+        relative_t = np.diff(np.insert(replay.t, 0, 0))
+
+        for i, keydown in enumerate(keydowns):
+            if keydown != 0:
+                # not sure why, but notelock can be reduced by 1ms if the keypress frame has a frametime of 0ms
+                # so we need to keep track of the frames where this happens
+                if relative_t[i] == 0:
+                    keydown_frames.append([replay.t[i], replay.xy[i], True])
+                else:
+                    keydown_frames.append([replay.t[i], replay.xy[i], False])
+
+        # add a duplicate frame when 2 keys are pressed at the same time
+        keydowns = keydowns[keydowns != 0]
+        i = 0
+        for j in np.where(keydowns == 3)[0]:
```

Comment:
why exactly 3? I assume this is `Key.M1 + Key.M2`, but this will ignore eg when both `Key.K1` and `Key.K2` are pressed. Shouldn't it be `keydowns & (Key.M1 + Key.M2) == (Key.M1 + Key.M2)`? (works for when all four bits are on, aka K1 and K2 are pressed, which are always set with M1 and M2 respectively - 15 & 3 == 3, and works for M1 and M2 as well, 3 & 3 == 3)

Reply:
I think `keypresses = replay.k & Investigator.MASK` already converts Key.K1 and Key.K2 to Key.M1 and Key.M2 respectively

Reply:
oh, well spotted. You're right. Let's still compare keydowns to MASK then - `keydowns == Investigator.MASK`, to remove the magic number 3